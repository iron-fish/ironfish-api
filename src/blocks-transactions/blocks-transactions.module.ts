/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { ApiConfigModule } from '../api-config/api-config.module';
import { EventsModule } from '../events/events.module';
import { PrismaModule } from '../prisma/prisma.module';
import { BlocksTransactionsService } from './blocks-transactions.service';

@Module({
  exports: [BlocksTransactionsService],
  imports: [ApiConfigModule, EventsModule, PrismaModule],
  providers: [BlocksTransactionsService],
})
export class BlocksTransactionsModule {}
