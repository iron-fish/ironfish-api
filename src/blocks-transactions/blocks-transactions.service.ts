/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable, UnprocessableEntityException } from '@nestjs/common';
import { Block, BlockTransaction, Transaction } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import { ListBlockTransactionOptions } from './interfaces/list-block-transactions-options';

@Injectable()
export class BlocksTransactionsService {
  constructor(private readonly prisma: PrismaService) {}

  async upsert(
    block: Block,
    transaction: Transaction,
  ): Promise<BlockTransaction> {
    return this.prisma.$transaction(async (prisma) => {
      return prisma.blockTransaction.upsert({
        create: {
          block_id: block.id,
          transaction_id: transaction.id,
        },
        update: {
          block_id: block.id,
          transaction_id: transaction.id,
        },
        where: {
          block_id_transaction_id: {
            block_id: block.id,
            transaction_id: transaction.id,
          },
        },
      });
    });
  }

  async list(
    options: ListBlockTransactionOptions,
  ): Promise<BlockTransaction[]> {
    if (options.blockId) {
      return this.prisma.blockTransaction.findMany({
        where: {
          block_id: options.blockId,
        },
      });
    } else if (options.transactionId) {
      return this.prisma.blockTransaction.findMany({
        where: {
          transaction_id: options.transactionId,
        },
      });
    } else {
      throw new UnprocessableEntityException();
    }
  }
}
