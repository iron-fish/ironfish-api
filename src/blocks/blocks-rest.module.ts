/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { UsersModule } from '../users/users.module';
import { BlocksController } from './blocks.controller';
import { BlocksModule } from './blocks.module';

@Module({
  controllers: [BlocksController],
  imports: [BlocksModule, UsersModule],
})
export class BlocksRestModule {}
