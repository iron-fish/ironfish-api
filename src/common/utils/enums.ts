/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

export type StrEnumValue<T> = T[keyof T];
export type StrEnum<T> = Record<keyof T, string>;

function getEnumStrValues<T extends StrEnum<T>>(
  enumType: T,
): Array<StrEnumValue<T>> {
  return Object.values(enumType)
    .filter((v) => typeof v === 'string')
    .map((v) => v as StrEnumValue<T>);
}

export function isInStrEnum<T extends StrEnum<T>>(
  value: unknown,
  enumType: T,
): boolean {
  return !!getEnumStrValues(enumType).find((enumValue) => enumValue === value);
}
