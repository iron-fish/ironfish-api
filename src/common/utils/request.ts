/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Request } from 'express';

export function fetchIpAddressFromRequest(request: Request): string {
  const xForwardedFor = request.header('X-Forwarded-For');
  if (xForwardedFor) {
    const addresses = xForwardedFor.split(',');
    if (addresses.length) {
      return addresses[0].trim();
    }
  }

  // In Heroku, `X-Forwarded-For` will always be set. However, this will be used
  // as a fallback in local development.
  return request.ip;
}
