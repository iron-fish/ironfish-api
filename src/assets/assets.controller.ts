/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  Controller,
  Get,
  HttpStatus,
  NotFoundException,
  Query,
  ValidationPipe,
} from '@nestjs/common';
import { ApiOperation, ApiTags } from '@nestjs/swagger';
import { BlocksTransactionsService } from '../blocks-transactions/blocks-transactions.service';
import { PaginatedList } from '../common/interfaces/paginated-list';
import { TransactionsService } from '../transactions/transactions.service';
import { AssetsService } from './assets.service';
import { AssetQueryDto } from './dto/asset-query-dto';
import { AssetsQueryDto } from './dto/assets-query.dto';
import { SerializedAsset } from './interfaces/serialized-asset';

@ApiTags('Assets')
@Controller('assets')
export class AssetsController {
  constructor(
    private readonly assetsService: AssetsService,
    private readonly blocksTransactionsService: BlocksTransactionsService,
    private readonly transactionsService: TransactionsService,
  ) {}

  @ApiOperation({ summary: 'Gets an asset by identifier' })
  @Get('find')
  async find(
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { id }: AssetQueryDto,
  ): Promise<SerializedAsset> {
    const asset = await this.assetsService.findByIdentifierOrThrow(id);
    const transaction = await this.transactionsService.findOrThrow(
      asset.created_transaction_id,
    );

    const blocks = await this.blocksTransactionsService.findBlocksByTransaction(
      transaction,
    );
    const block = blocks.find((block) => block.main);
    if (!block) {
      throw new NotFoundException();
    }

    return {
      object: 'asset',
      created_transaction_hash: transaction.hash,
      created_transaction_timestamp: block.timestamp.toISOString(),
      id: asset.id,
      identifier: asset.identifier,
      metadata: asset.metadata,
      name: asset.name,
      owner: asset.owner,
      supply: asset.supply.toString(),
    };
  }

  @ApiOperation({ summary: 'Lists assets' })
  @Get()
  async list(
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    dto: AssetsQueryDto,
  ): Promise<PaginatedList<SerializedAsset>> {
    const { data, hasNext, hasPrevious } = await this.assetsService.list(dto);

    const serializedData: SerializedAsset[] = [];
    for (const asset of data) {
      const transaction = await this.transactionsService.findOrThrow(
        asset.created_transaction_id,
      );

      const blocks =
        await this.blocksTransactionsService.findBlocksByTransaction(
          transaction,
        );
      const block = blocks.find((block) => block.main);
      if (!block) {
        continue;
      }

      serializedData.push({
        object: 'asset',
        created_transaction_hash: transaction.hash,
        created_transaction_timestamp: block.timestamp.toISOString(),
        id: asset.id,
        identifier: asset.identifier,
        metadata: asset.metadata,
        name: asset.name,
        owner: asset.owner,
        supply: asset.supply.toString(),
      });
    }

    return {
      object: 'list',
      data: serializedData,
      metadata: {
        has_next: hasNext,
        has_previous: hasPrevious,
      },
    };
  }
}
