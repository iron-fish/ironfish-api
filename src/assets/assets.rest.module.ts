/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { BlocksTransactionsModule } from '../blocks-transactions/blocks-transactions.module';
import { TransactionsModule } from '../transactions/transactions.module';
import { AssetsController } from './assets.controller';
import { AssetsModule } from './assets.module';

@Module({
  controllers: [AssetsController],
  imports: [AssetsModule, BlocksTransactionsModule, TransactionsModule],
})
export class AssetsRestModule {}
