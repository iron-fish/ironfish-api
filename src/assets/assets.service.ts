/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable, NotFoundException } from '@nestjs/common';
import { DEFAULT_LIMIT, MAX_LIMIT } from '../common/constants';
import { SortOrder } from '../common/enums/sort-order';
import { PrismaService } from '../prisma/prisma.service';
import { BasePrismaClient } from '../prisma/types/base-prisma-client';
import { CreateAssetOptions } from './interfaces/create-asset-options';
import { ListAssetsOptions } from './interfaces/list-assets-optionts';
import { Asset, Prisma, Transaction } from '.prisma/client';

@Injectable()
export class AssetsService {
  constructor(private readonly prisma: PrismaService) {}

  async findOrThrow(id: number): Promise<Asset> {
    return this.findOrThrowWithClient(id, this.prisma.readClient);
  }

  async findOrThrowWithClient(
    id: number,
    prisma: BasePrismaClient,
  ): Promise<Asset> {
    const record = await prisma.asset.findUnique({
      where: {
        id,
      },
    });
    if (!record) {
      throw new NotFoundException();
    }
    return record;
  }

  async findByIdentifierOrThrow(identifier: string): Promise<Asset> {
    return this.findByIdentifierOrThrowWithClient(identifier, this.prisma);
  }

  async findByIdentifierOrThrowWithClient(
    identifier: string,
    prisma: BasePrismaClient,
  ): Promise<Asset> {
    const record = await prisma.asset.findUnique({
      where: {
        identifier,
      },
    });
    if (!record) {
      throw new NotFoundException();
    }
    return record;
  }

  async upsert(
    options: CreateAssetOptions,
    transaction: Transaction,
    prisma: BasePrismaClient,
  ): Promise<Asset> {
    return prisma.asset.upsert({
      create: {
        identifier: options.identifier,
        metadata: options.metadata,
        name: options.name,
        owner: options.owner,
        supply: BigInt(0),
        created_transaction_id: transaction.id,
      },
      update: {},
      where: {
        identifier: options.identifier,
      },
    });
  }

  async updateSupply(
    asset: Asset,
    delta: bigint,
    prisma: BasePrismaClient,
  ): Promise<Asset> {
    return prisma.asset.update({
      data: {
        supply: {
          increment: delta,
        },
      },
      where: {
        id: asset.id,
      },
    });
  }

  async list(options: ListAssetsOptions): Promise<{
    data: Asset[];
    hasNext: boolean;
    hasPrevious: boolean;
  }> {
    const cursorId = options.before ?? options.after;
    const cursor = cursorId ? { id: cursorId } : undefined;
    const direction = options.before !== undefined ? -1 : 1;
    const limit =
      direction * Math.min(MAX_LIMIT, options.limit || DEFAULT_LIMIT);
    const orderBy = { id: SortOrder.DESC };
    const skip = cursor ? 1 : 0;

    let where: Prisma.AssetWhereInput = {};
    if (options.search) {
      where = {
        name: {
          contains: options.search,
        },
      };
    }

    const data = await this.prisma.readClient.asset.findMany({
      cursor,
      orderBy,
      skip,
      take: limit,
      where,
    });

    return {
      data,
      ...(await this.getListMetadata(data, where, orderBy)),
    };
  }

  private async getListMetadata(
    data: Asset[],
    where: Prisma.AssetWhereInput,
    orderBy: Prisma.Enumerable<Prisma.AssetOrderByWithRelationInput>,
  ): Promise<{ hasNext: boolean; hasPrevious: boolean }> {
    const { length } = data;
    if (length === 0) {
      return {
        hasNext: false,
        hasPrevious: false,
      };
    }
    const nextRecords = await this.prisma.readClient.asset.findMany({
      where,
      orderBy,
      cursor: { id: data[length - 1].id },
      skip: 1,
      take: 1,
    });
    const previousRecords = await this.prisma.readClient.asset.findMany({
      where,
      orderBy,
      cursor: { id: data[0].id },
      skip: 1,
      take: -1,
    });
    return {
      hasNext: nextRecords.length > 0,
      hasPrevious: previousRecords.length > 0,
    };
  }
}
