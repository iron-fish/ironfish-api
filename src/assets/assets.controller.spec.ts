/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { HttpStatus, INestApplication } from '@nestjs/common';
import faker from 'faker';
import request from 'supertest';
import { v4 as uuid } from 'uuid';
import { BlocksService } from '../blocks/blocks.service';
import { BlockOperation } from '../blocks/enums/block-operation';
import { BlocksTransactionsService } from '../blocks-transactions/blocks-transactions.service';
import { PrismaService } from '../prisma/prisma.service';
import { bootstrapTestApp } from '../test/test-app';
import { TransactionsService } from '../transactions/transactions.service';
import { AssetsService } from './assets.service';

describe('AssetsController', () => {
  let app: INestApplication;
  let assetsService: AssetsService;
  let blocksService: BlocksService;
  let blocksTransactionsService: BlocksTransactionsService;
  let prisma: PrismaService;
  let transactionsService: TransactionsService;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    assetsService = app.get(AssetsService);
    blocksService = app.get(BlocksService);
    blocksTransactionsService = app.get(BlocksTransactionsService);
    prisma = app.get(PrismaService);
    transactionsService = app.get(TransactionsService);
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('GET /assets/find', () => {
    describe('with an invalid query', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/assets/find')
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a missing record', () => {
      it('returns a 404', async () => {
        await request(app.getHttpServer())
          .get('/assets/find')
          .query({ id: 'asdf' })
          .expect(HttpStatus.NOT_FOUND);
      });
    });

    describe('when the asset has not been added to a block', () => {
      it('returns a 404', async () => {
        const transaction = (
          await transactionsService.createMany([
            {
              fee: 0,
              hash: uuid(),
              notes: [],
              size: 0,
              spends: [],
            },
          ])
        )[0];

        const asset = await assetsService.upsert(
          {
            identifier: uuid(),
            metadata: uuid(),
            name: uuid(),
            owner: uuid(),
          },
          transaction,
          prisma,
        );

        await request(app.getHttpServer())
          .get('/assets/find')
          .query({ id: asset.identifier })
          .expect(HttpStatus.NOT_FOUND);
      });
    });

    describe('with a valid identifier', () => {
      it('returns the asset', async () => {
        const { block } = await blocksService.upsert(prisma, {
          hash: uuid(),
          sequence: faker.datatype.number(),
          difficulty: faker.datatype.number(),
          work: faker.datatype.number(),
          timestamp: new Date(),
          transactionsCount: 1,
          type: BlockOperation.CONNECTED,
          graffiti: uuid(),
          previousBlockHash: uuid(),
          size: faker.datatype.number(),
        });
        const transaction = (
          await transactionsService.createMany([
            {
              fee: 0,
              hash: uuid(),
              notes: [],
              size: 0,
              spends: [],
            },
          ])
        )[0];
        await blocksTransactionsService.upsert(prisma, block, transaction, 0);

        const asset = await assetsService.upsert(
          {
            identifier: uuid(),
            metadata: uuid(),
            name: uuid(),
            owner: uuid(),
          },
          transaction,
          prisma,
        );

        const { body } = await request(app.getHttpServer())
          .get('/assets/find')
          .query({ id: asset.identifier })
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          created_transaction_hash: transaction.hash,
          created_transaction_timestamp: block.timestamp.toISOString(),
          identifier: asset.identifier,
          metadata: asset.metadata,
          name: asset.name,
          owner: asset.owner,
          supply: asset.supply.toString(),
        });
      });
    });
  });

  describe('GET /assets', () => {
    it('returns the assets', async () => {
      const { block } = await blocksService.upsert(prisma, {
        hash: uuid(),
        sequence: faker.datatype.number(),
        difficulty: faker.datatype.number(),
        work: faker.datatype.number(),
        timestamp: new Date(),
        transactionsCount: 1,
        type: BlockOperation.CONNECTED,
        graffiti: uuid(),
        previousBlockHash: uuid(),
        size: faker.datatype.number(),
      });
      const transaction = (
        await transactionsService.createMany([
          {
            fee: 0,
            hash: uuid(),
            notes: [],
            size: 0,
            spends: [],
          },
        ])
      )[0];
      await blocksTransactionsService.upsert(prisma, block, transaction, 0);

      const asset = await assetsService.upsert(
        {
          identifier: uuid(),
          metadata: uuid(),
          name: uuid(),
          owner: uuid(),
        },
        transaction,
        prisma,
      );

      const secondTransaction = (
        await transactionsService.createMany([
          {
            fee: 0,
            hash: 'foo',
            notes: [],
            size: 0,
            spends: [],
          },
        ])
      )[0];
      await blocksTransactionsService.upsert(
        prisma,
        block,
        secondTransaction,
        1,
      );
      const secondAsset = await assetsService.upsert(
        {
          identifier: uuid(),
          metadata: uuid(),
          name: uuid(),
          owner: uuid(),
        },
        secondTransaction,
        prisma,
      );

      const { body } = await request(app.getHttpServer())
        .get('/assets')
        .query({ limit: 2 })
        .expect(HttpStatus.OK);

      expect(body).toMatchObject({
        object: 'list',
        data: [
          {
            object: 'asset',
            created_transaction_hash: secondTransaction.hash,
            created_transaction_timestamp: block.timestamp.toISOString(),
            id: secondAsset.id,
            identifier: secondAsset.identifier,
            metadata: secondAsset.metadata,
            name: secondAsset.name,
            owner: secondAsset.owner,
            supply: secondAsset.supply.toString(),
          },
          {
            object: 'asset',
            created_transaction_hash: transaction.hash,
            created_transaction_timestamp: block.timestamp.toISOString(),
            id: asset.id,
            identifier: asset.identifier,
            metadata: asset.metadata,
            name: asset.name,
            owner: asset.owner,
            supply: asset.supply.toString(),
          },
        ],
        metadata: {
          has_next: true,
          has_previous: false,
        },
      });
    });
  });
});
