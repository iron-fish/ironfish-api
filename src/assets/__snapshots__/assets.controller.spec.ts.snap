// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetsController GET /assets/find with an invalid query returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "\\"id\\" required to query for asset",
    "id must be a string",
  ],
  "statusCode": 422,
}
`;
